### PR TITLE
Backend ChatChannelMembership Search using Elasticsearch

### PR DESCRIPTION
**Pull request by mstruve**
## What type of PR is this? (check all applicable)
- [x] Feature

## Description
This is the backend code that we will use when searching Elasticsearch for ChatChannelMemberships. Similar to how we set up tags, the next PR will create a search controller chat_channel_memberships action that will take a set of params and the current_user and run a search using them.
```
Search::ChatChannelMembership.search_documents(params: params, user_id: current_user.id)
```

I choose to separate the frontend from the backend since all in one PR it is a lot of code and I like being able to focus on one at a time. I also think it is easier for others to review that way. 

## Related Tickets & Documents
https://github.com/thepracticaldev/dev.to/projects/6#card-32289262

## Added tests?
- [x] yes

## Added to documentation?
- [x] no documentation needed

alt_text

### CODE REVIEW

File: app/services/search/query_builders/chat_channel_membership.rb [hunk 1]
```diff
@@ -0,0 +1,77 @@
+module Search
+  module QueryBuilders
+    class ChatChannelMembership
+      FILTER_KEYS = %i[
```
Comment (mstruve) [inline app/services/search/query_builders/chat_channel_membership.rb]: These are all keys that [we currently use as filters](https://github.com/thepracticaldev/dev.to/blob/master/app/models/chat_channel_membership.rb#L25), I set them up in an array so we can easily add more when we want to.

File: app/services/search/query_builders/chat_channel_membership.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        channel_status
+        channel_type
+        status
+        viewable_by
+      ].freeze
+
+      QUERY_KEYS = %i[
```
Comment (mstruve) [inline app/services/search/query_builders/chat_channel_membership.rb]: These are fields we want to run text searches on, ie score documents based on how well they match. For more information on queries vs filters [checkout the Elasticsearch docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-filter-context.html)

File: spec/features/search/chat_channel_membership_spec.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
+require "rails_helper"
+
+RSpec.describe Search::ChatChannelMembership, type: :feature, elasticsearch: true do
```
Comment (mstruve) [inline spec/features/search/chat_channel_membership_spec.rb]: I really wanted solid end-to-end search specs bc in the past I have found these are BY FAR the most useful at catching bugs when adding new code. Yes, they are slower but trust me, we want these. I choose to make a new feature folder for them. If someone has another suggested approach let me know.

**Comment by atsmith813** [comment]
Is the game plan to re-factor this after a few iterations? Or that the pattern going forward is to create a `QueryBuilder` for each model?

**Comment by mstruve** [comment]
Moving forward I would say start with a query builder for each model and then see where it gets us. Stepping back I do think this will be similar to the Search classes where we can extract a bunch of shared logic like building filters and building queries. 

I'm interested to see how "searchables" (Article, User, PodcastEpisode) plays out since that is the first set of docs where the model is not directly correlated to an ES doc type. At least that is how I think its going to work.

File: app/services/search/query_builders/chat_channel_membership.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        channel_text
+      ].freeze
+
+      DEFAULT_PARAMS = {
+        sort_by: "channel_last_message_at",
+        sort_direction: "desc",
+        size: 0
```
Comment (rhymes) [inline app/services/search/query_builders/chat_channel_membership.rb]: what's size?

File: spec/features/search/chat_channel_membership_spec.rb (hunk 3 above)
Comment (rhymes) [reply spec/features/search/chat_channel_membership_spec.rb]: I'm a little bit confused about this choice, in the sense that for RSpec feature specs are basically system tests designed to excercise the app from the outside: https://relishapp.com/rspec/rspec-rails/docs/feature-specs/feature-spec

but these are normal services specs.

What about just moving this in `spec/services/search`? You can either add them to `chat_channel_membership_spec.rb` in a `describe .search_documents` section or add a separate file like `chat_channel_membership_search_documents.rb` or `spec/services/search/chat_channel_membership/search_documents.rb`, your call :)

File: app/services/search/query_builders/chat_channel_membership.rb (hunk 4 above)
Comment (mstruve) [reply app/services/search/query_builders/chat_channel_membership.rb]: How many documents are returned. Elasticsearch by default returns 10, I wanted to make our default size 0 to make sure people are explicit about requesting the correct number.

File: spec/features/search/chat_channel_membership_spec.rb (hunk 3 above)
Comment (mstruve) [reply spec/features/search/chat_channel_membership_spec.rb]: I thought of them as feature specs bc they aren't just testing the service's ability to talk to Elasticsearch they are also testing the query builders ability to form a valid query and then that the query returns valid results. It felt a little weird bc if we add another filter key to query builder, this is where we would want to add another spec which is what led me to consider them "feature" specs as opposed to service specs but I am game for consolidating!

File: app/services/search/query_builders/chat_channel_membership.rb (hunk 4 above)
Comment (rhymes) [reply app/services/search/query_builders/chat_channel_membership.rb]: Got it! +1 for adding this to the inline documentation of the search method :D As, according to:

https://github.com/thepracticaldev/dev.to/pull/6217/files/573a6f6044d0a76ea8b19ad504937c8230c97df4#diff-c8d5692235cc1a81cb7ed64d40688e2cR53

it won't break, just do a HTTP call that will always return 0 results

File: spec/features/search/chat_channel_membership_spec.rb (hunk 3 above)
Comment (rhymes) [reply spec/features/search/chat_channel_membership_spec.rb]: Got it! +1 for consolidating. Most service specs test features anyway. I think it's just a matter of naming, as what we call "feature spec" is not what RSpec calls "feature spec" :D

**Pull request merged**